git-pr-merge

```diff
@@ -65,40 +87,114 @@ prmerge::setup() {
   fi
 }
 
-prmerge::prepare() {
-  echo merging "${feature}" to "${master}"
-
-  # Fetch from all remotes so we can check if local branch is up-to-date
-  git remote update >/dev/null
-
-  if [[ "$(git rev-parse "${master}")" != "$(git rev-parse "${master}@{upstream}")" ]]; then
-    printf '%s is not up-to-date. You should run:\n' "${master}" >&2
-    printf '  git checkout %s\n' "${master}" >&2
-    printf '  git pull\n' >&2
-    return 1
-  fi
-
+common::prepare() {
   # cd to root of repo as master may not contain CWD, but switch back
   # when we're done.
   # shellcheck disable=SC2064
   # We want to evaluate $(pwd) now and not later.
   trap "cd \"$(pwd)\"" EXIT
   cd "$(git rev-parse --show-toplevel)"
 
+  # Fetch from all remotes so we can check if local branch is up-to-date
+  git remote update >/dev/null
+}
+
+prmerge::prepare() {
+  common::prepare
+
+  local do_pull=false
+  if [[ "$(git rev-parse "${master}")" != "$(git rev-parse "${master}@{upstream}")" ]]; then
+    printf '\n%s is not up-to-date. To update, press enter (^C to abort):' "${master}"
+    read -r
+    do_pull=true
+  fi
+
   git checkout -q "${master}"
+
+  if "${do_pull}"; then
+    git pull
+    if ! git merge-base --is-ancestor "${master}" "${feature}"; then
+      git checkout "${feature}"
+      printf 'Local %s is now up-to-date. Update %s before merging\n' "${master}" "${feature}"
+      exit 1
+    fi
+  fi
+}
+
+prupdate::prepare() {
+  common::prepare
+
+  if git merge-base --is-ancestor "${master}@{upstream}" "${feature}"; then
+    printf 'Already up to date\n'
+    exit 0
+  fi
+
+  # We need to ensure the PR base.sha is up-to-date, otherwise we create a
+  # merge commit on the PR that dismisses approval. This can be done by
+  # force pushing a commit on the PR feature branch, but we need to make sure
+  # we use a push/commit that does not remove approval.
```

Comment:
... a merge commit of master into feature meets this criterion.

Reply:
i'm not sure it does. I may have misremembered, but ISTR that the force push is what updated the base.sha

Reply:
I'm not sure either, if you feel confident with it, I'm happy.

Reply:
yeah, i'm not sure it does not ;)
it seems to simulate what github does when clicking "update branch' button - but maybe I'm mistaken.